### PR TITLE
Enable unit tests for native driver

### LINKED ISSUE #775: Add tests for bpf2c_driver.c and bpf2c_dll.c
There should be tests for [bpf2c_driver.c](https://github.com/microsoft/ebpf-for-windows/blob/master/tools/bpf2c/bpf2c_driver.c) and [bpf2c_dll.c](https://github.com/microsoft/ebpf-for-windows/blob/master/tools/bpf2c/bpf2c_dll.c).

### PR DESCRIPTION
**Pull request by saxena-anurag**
## Description

1. This PR enables applicable unit tests for native driver load scenario.
2. Also fixes some bugs in native module, test code, etc.

## Testing

Added test cases cover the changes.

## Documentation

NA

Fixes #775

### CODE REVIEW

**Comment by Alan-Jowett** [comment]
Looks like bpf2c output changed. Can you please regenerate the affected expected results?

File: tests/end_to_end/test_helper.cpp [hunk 1]
```diff
@@ -230,13 +232,18 @@ _unload_all_native_modules()
 static void
 _load_native_module(_Inout_ service_context_t* context)
 {
-    context->dll = LoadLibraryW(context->file_path.c_str());
+    // For user mode tests, all the sample programs are compiled into a single DLL file.
+    context->dll = LoadLibraryA(NATIVE_DLL_NAME);
+    // context->dll = LoadLibraryW(context->file_path.c_str());
```
Comment (dthaler) [inline tests/end_to_end/test_helper.cpp:237]: remove commented out code

Comment (saxena-anurag) [reply tests/end_to_end/test_helper.cpp]: Done.

**Review by Alan-Jowett — APPROVED** [review]

**Review by Alan-Jowett — APPROVED** [review]

File: tests/end_to_end/test_helper.cpp [hunk 2]
```diff
@@ -230,13 +238,29 @@ _unload_all_native_modules()
...
+    // For user mode tests, most of the sample programs are compiled into a single DLL file.
+    // First try loading from there.
+    context->dll = LoadLibraryA(NATIVE_DLL_NAME);
     REQUIRE(context->dll != nullptr);
 
-    auto get_function =
-        reinterpret_cast<decltype(&get_metadata_table)>(GetProcAddress(context->dll, "get_metadata_table"));
-    REQUIRE(get_function != nullptr);
-    metadata_table_t* table = get_function();
+    // Get file name from the file path
```
Comment (dthaler) [inline tests/end_to_end/test_helper.cpp:246]:
```suggestion
    // Get file name from the file path.
```

File: tests/end_to_end/end_to_end.cpp [hunk 3]
```diff
@@ -607,13 +623,22 @@ bindmonitor_tailcall_test(ebpf_execution_type_t execution_type)
     index = 1;
     REQUIRE(bpf_map_update_elem(prog_map_fd, &index, &callee1_fd, 0) == 0);
 
-    // Test map-in-maps.
+    // Validate various maps.
+
+    // Validate map-in-maps with "inner_id"
```
Comment (dthaler) [inline tests/end_to_end/end_to_end.cpp:628]:
```suggestion
    // Validate map-in-maps with "inner_id".
```

File: tests/end_to_end/end_to_end.cpp [hunk 4]
```diff
@@ -607,13 +623,22 @@ bindmonitor_tailcall_test(ebpf_execution_type_t execution_type)
...
     struct bpf_map* outer_map = bpf_object__find_map_by_name(object, "dummy_outer_map");
     REQUIRE(outer_map != nullptr);
 
     int outer_map_fd = bpf_map__fd(outer_map);
     REQUIRE(outer_map_fd > 0);
 
+    // Validate map-in-maps with "inner_idx"
```
Comment (dthaler) [inline tests/end_to_end/end_to_end.cpp:635]:
```suggestion
    // Validate map-in-maps with "inner_idx".
```

**Review by dthaler — APPROVED** [review]

**Pull request merged**